pest/src/parser_state.rs

```diff
@@ -1149,4 +1157,81 @@ mod test {
         assert_eq!(normalize_index(-5, 5), Some(0));
         assert_eq!(normalize_index(-6, 3), None);
     }
+
+    #[test]
+    fn test_restoring_on_err() {
+        // This corresponds roughly to Pest syntax:
+        //
+        // stack_use = @{ PUSH("A")+ ~ "B" ~ POP_ALL }
```

Comment:
I would make this part of https://github.com/pest-parser/pest/blob/master/derive/tests/grammar.pest#L38 in order not to have to define this manually.